packages/core/contracts/financial-templates/optimistic-distributor/OptimisticDistributor.sol

```diff
@@ -0,0 +1,425 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+pragma solidity ^0.8.0;
+
+import "../../common/implementation/AncillaryData.sol";
+import "../../common/implementation/Lockable.sol";
+import "../../common/implementation/MultiCaller.sol";
+import "../../common/implementation/Testable.sol";
+import "../../common/interfaces/AddressWhitelistInterface.sol";
+import "../../merkle-distributor/implementation/MerkleDistributor.sol";
+import "../../oracle/implementation/Constants.sol";
+import "../../oracle/interfaces/FinderInterface.sol";
+import "../../oracle/interfaces/IdentifierWhitelistInterface.sol";
+import "../../oracle/interfaces/OptimisticOracleInterface.sol";
+import "../../oracle/interfaces/StoreInterface.sol";
+
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+
+/**
+ * @title  OptimisticDistributor contract.
+ * @notice Allows sponsors to distribute rewards through MerkleDistributor contract secured by UMA Optimistic Oracle.
+ */
+contract OptimisticDistributor is Lockable, MultiCaller, Testable {
+    using SafeERC20 for IERC20;
+
+    /********************************************
+     *  OPTIMISTIC DISTRIBUTOR DATA STRUCTURES  *
+     ********************************************/
+
+    // Represents reward posted by a sponsor.
+    struct Reward {
+        address sponsor;
+        IERC20 rewardToken;
+        uint256 maximumRewardAmount;
+        uint256 earliestProposalTimestamp;
+        uint256 optimisticOracleProposerBond;
+        uint256 optimisticOracleLivenessTime;
+        bytes32 priceIdentifier;
+        bytes customAncillaryData;
+    }
+
+    // Represents proposed rewards distribution.
+    struct Proposal {
+        uint256 rewardIndex;
+        uint256 timestamp;
+        bytes32 merkleRoot;
+        string ipfsHash;
+    }
+
+    /********************************************
+     *      STATE VARIABLES AND CONSTANTS       *
+     ********************************************/
+
+    // Reserve for bytes appended to ancillary data (e.g. OracleSpoke) when resolving price from non-mainnet chains.
+    // This also covers appending proposalIndex by this contract.
+    uint256 public constant ANCILLARY_BYTES_RESERVE = 512;
+
+    // Restrict Optimistic Oracle liveness to between 10 minutes and 100 years.
+    uint256 public constant MINIMUM_LIVENESS = 10 minutes;
+    uint256 public constant MAXIMUM_LIVENESS = 5200 weeks;
+
+    // Final fee can be synced and stored in the contract.
+    uint256 public finalFee;
+
+    // Ancillary data length limit can be synced and stored in the contract.
+    uint256 public ancillaryBytesLimit;
+
+    // Index of next created reward or proposal.
+    uint256 public nextCreatedReward;
+    uint256 public nextCreatedProposal;
+
+    // Rewards and proposals are mapped to their indices.
+    mapping(uint256 => Reward) public rewards;
```

Comment:
Why not just use an array so we don't need to store `nextCreatedReward` and `nextCreatedProposal`? @chrismaree wdyt

Reply:
At some point I was considering addressing struct mappings by hashing some of its fields as index, thus allowing to detect duplicate proposals as raised [here](https://github.com/UMAprotocol/protocol/pull/3861/files#r848957341). In case we address this by allowing only 1 concurrent proposal per reward (with enum state element to reward struct) I indeed should use dynamic array instead.